napi_js_data.md

```diff
@@ -0,0 +1,760 @@
+# N-API - JavaScript Data Types and Values 
+
+N-API exposes a set of APIs to create fundamental JavaScript Objects. 
+Some of these types are documented under [Section 6](https://tc39.github.io/ecma262/#sec-ecmascript-data-types-and-values) 
+of the [ECMAScript Language Specification](https://tc39.github.io/ecma262/).
+
+Fundamentally, these APIs are used to do one of the following:
+1. Create a new JavaScript object
+2. Convert from a primitive C++ type to N-API value
+3. Convert from N-API value to a primitive C++ type
+
+N-API values are represented by the type `napi_value`.
+Any N-API API that requires a JavaScript object takes in a `napi_value`.
+In some cases, the API does check the type of the `napi_value` up-front.
+However, for better performance, it's better that the caller makes sure that 
+the `napi_value` in question is of the JavaScript type expected by the API.
+
+## Enum types
+
+### *napi_value_type*
+
+#### Definition
+```
+enum napi_valuetype {
+  napi_undefined,
+  napi_null,
+  napi_boolean,
+  napi_number,
+  napi_string,
+  napi_symbol,
+  napi_object,
+  napi_function,
+  napi_external,
+}
+```
+
+#### Description
+Describes the type of a napi_value. This generally corresponds with the types described
+in [Section 6.1 of the ECMAScript Language Specification](https://tc39.github.io/ecma262/#sec-ecmascript-language-types).
+In addition to types in that section, `napi_value_type` also can represent Functions and Objects with external data.
+
+### *napi_typedarray_type*
+
+#### Definition
+```
+enum napi_typedarray_type {
+  napi_int8_array,
+  napi_uint8_array,
+  napi_uint8_clamped_array,
+  napi_int16_array,
+  napi_uint16_array,
+  napi_int32_array,
+  napi_uint32_array,
+  napi_float32_array,
+  napi_float6_array4
+}
+```
+
+#### Description
+This represents the underlying binary scalar datatype of the TypedArray.
+Elements of this enum correspond to [Section 22.2 of the ECMAScript Language Specification](https://tc39.github.io/ecma262/#sec-typedarray-objects).
+
+## Functions
+
+### *napi_create_array*
+
+#### Signature
+```
+napi_status napi_create_array(napi_env env, napi_value* result)
+```
+
+#### Parameters
+- `[in]  env`: The environment that the N-API API is invoked under
+- `[out] result`: A `napi_value` representing a JavaScript Array
+
+#### Return value
+- `napi_ok` if the API succeeded.
```

Comment:
Do we need to document the `napi_ok` return value everywhere? It seems pointless when not also documenting the possible error return values. Probably we should just have one topic with documentation of all the return status codes.